tests/Integration/Behaviour/Features/Context/Domain/CartRule/AbstractCartRuleFeatureContext.php

```diff
@@ -28,13 +28,275 @@
 
 namespace Tests\Integration\Behaviour\Features\Context\Domain\CartRule;
 
+use DateTime;
+use PHPUnit\Framework\Assert;
+use PrestaShop\Decimal\DecimalNumber;
+use PrestaShop\PrestaShop\Core\Domain\CartRule\Command\AddCartRuleCommand;
+use PrestaShop\PrestaShop\Core\Domain\CartRule\Query\GetCartRuleForEditing;
+use PrestaShop\PrestaShop\Core\Domain\CartRule\QueryResult\CartRuleForEditing;
+use PrestaShop\PrestaShop\Core\Domain\CartRule\ValueObject\CartRuleId;
 use PrestaShop\PrestaShop\Core\Domain\ValueObject\Reduction;
 use PrestaShop\PrestaShop\Core\Form\IdentifiableObject\Builder\CartRule\CartRuleActionBuilder;
+use PrestaShop\PrestaShop\Core\Util\DateTime\DateTime as DateTimeUtil;
+use RuntimeException;
 use Tests\Integration\Behaviour\Features\Context\Domain\AbstractDomainFeatureContext;
 use Tests\Integration\Behaviour\Features\Context\Util\PrimitiveUtils;
+use Tests\Resources\DatabaseDump;
 
 abstract class AbstractCartRuleFeatureContext extends AbstractDomainFeatureContext
 {
+    /**
+     * @BeforeScenario @restore-cart-rules-before-scenario
```

Comment:
Publis static methods related to tags shouldn't be put in abstract classes, we did this in the past and it results in the method being called multiple time, which at best can decrease the performance as it's not optimized

But sometimes it also creates unexpected side effects which is more troublesome as it's hard to debug, the clear method such as these should be placed in a single independent Context or at least one that is only associated once to the test suite Maybe in `CartRuleAssertionFeatureContext` ?